lib/plugins/place_block.js

```diff
@@ -0,0 +1,97 @@
+const assert = require('assert')
+const { once } = require('events')
+const { callbackify } = require('../promise_utils')
+
+module.exports = inject
+
+function inject (bot, { version }) {
+  const Item = require('prismarine-item')(version)
+
+  async function placeBlockWithOptions (referenceBlock, faceVector, options) {
+    if (!bot.heldItem) throw new Error('must be holding an item to place a block')
+    // Look at the center of the face
+    const dx = 0.5 + faceVector.x * 0.5
+    let dy = 0.5 + faceVector.y * 0.5
+    const dz = 0.5 + faceVector.z * 0.5
+    if (dy === 0.5) {
+      if (options.half === 'top') dy += 0.25
+      else if (options.half === 'bottom') dy -= 0.25
+    }
+    await bot.lookAt(referenceBlock.position.offset(dx, dy, dz), false)
+    // TODO: tell the server that we are sneaking while doing this
+    bot.swingArm()
+    const pos = referenceBlock.position
+
+    if (bot.supportFeature('blockPlaceHasHeldItem')) {
+      bot._client.write('block_place', {
+        location: pos,
+        direction: vectorToDirection(faceVector),
+        heldItem: Item.toNotch(bot.heldItem),
+        cursorX: Math.floor(dx * 16),
+        cursorY: Math.floor(dy * 16),
+        cursorZ: Math.floor(dz * 16)
+      })
+    } else if (bot.supportFeature('blockPlaceHasHandAndIntCursor')) {
+      bot._client.write('block_place', {
+        location: pos,
+        direction: vectorToDirection(faceVector),
+        hand: 0,
+        cursorX: Math.floor(dx * 16),
+        cursorY: Math.floor(dy * 16),
+        cursorZ: Math.floor(dz * 16)
+      })
+    } else if (bot.supportFeature('blockPlaceHasHandAndFloatCursor')) {
+      bot._client.write('block_place', {
+        location: pos,
+        direction: vectorToDirection(faceVector),
+        hand: 0,
+        cursorX: dx,
+        cursorY: dy,
+        cursorZ: dz
+      })
+    } else if (bot.supportFeature('blockPlaceHasInsideBlock')) {
+      bot._client.write('block_place', {
+        location: pos,
+        direction: vectorToDirection(faceVector),
+        hand: 0,
+        cursorX: dx,
+        cursorY: dy,
+        cursorZ: dz,
+        insideBlock: false
+      })
+    }
+
+    const dest = pos.plus(faceVector)
+    const eventName = `blockUpdate:${dest}`
+
+    const [oldBlock, newBlock] = await once(bot, eventName)
+
+    if (oldBlock.type === newBlock.type) {
+      throw new Error(`No block has been placed : the block is still ${oldBlock.name}`)
+    }
+  }
+
+  async function placeBlock (referenceBlock, faceVector) {
+    await placeBlockWithOptions(referenceBlock, faceVector, {})
+  }
+
+  bot.placeBlock = callbackify(placeBlock)
+  bot._placeBlockWithOptions = placeBlockWithOptions
```

Comment:
What's the plan regarding this semi exposed function ?

Reply:
I'm not sure about the api yet, but this feature is needed. I thought we could have it as part of the unstable api for a while and see how it goes.